### ISSUE TITLE
Default Policy Parameters for Classifier

### ISSUE THREAD

**Issue by j0sh** [livepeer/verification-classifier#61]
In https://github.com/livepeer/research/issues/37 , the idea of policy-based verification was introduced. A local verification policy defines the configurable parameters of the verification algorithm. A broadcaster can use its policy to tune the behavior of the verifier, such as retrying the transcode with someone else if the calculated score is below a certain threshold, or adjusting the verification sampling rate of based on its tolerances for risk, latency or computation expense.

The policy parameters for this classifier may include:

* Acceptable threshold for the continuous score
* The sampling rate (of entire segments)
* Number of retries

It would be good to incorporate a reasonable default values for these parameters (and any others parameters that may be relevant). The question here is what constitutes reasonable defaults? Some things that would be good to know when selecting these values:

* Confidence interval for a given threshold
* How the selection of parameters impacts the likelihood of false negatives or false positives 
* Sensitivity of these values to the type of content that's being verified

(For retries, we can assume that a different encoder is used which yield a different scores, and there is no collusion between transcoders.)

**Comment by yondonfu** [comment]
> The policy parameters for this classifier may include

Are you suggesting that the policy parameters are implemented on the verifier itself? I was actually thinking that the broadcaster could implement the policy parameters and the verifier is only responsible for responding to verification requests with the logic of when to send verification requests implemented by the broadcaster.

> Acceptable threshold for the continuous score

What starting with 0? The verifier API currently returns a tamper score (via the `tamper` field in the response) that will be negative if the result is classified as tamper and a lower score should represent a higher certainty that the result was tampered. So, a threshold set to 0 would mean that all video that is not classified as tampered would be accepted. If N results are below 0 (where N is the max # of retries), we pick the highest scoring result meaning we pick the result that is the least distorted and the least likely to be tampered [1].

[1] If we're assuming different encoders yielding different scores and no collusion

> The sampling rate (of entire segments)

Perhaps we can start with a default of 100% in a first pass and then adjust after observing some e2e performance?

**Comment by j0sh** [comment]
>  I was actually thinking that the broadcaster could implement the policy parameters

Agreed that the enforcement of the policy belongs on the broadcaster. But it'd be good to know what appropriate default values are for this classifier, since the policy for a given verifier is likely to be tuned to the characteristics of the verifier itself.

> What about starting with 0?

@ndujar How does that sound? What would be the effects of a threshold that is higher or lower, given the current behavior of the classifier? I'm curious as to the degree this affects the retry rate. More generally, I'm curious about the distribution of scores, especially for videos at risk for false negatives or false positives, how close are they to the boundary?

Eg, one thought is setting a slightly higher threshold may filter out false negatives (tampered video that's flagged as non-tampered) if we're reasonably certain that non-tampered video obtained through a retry would be comfortably above this threshold.

> If N results are below 0

What's a good number for N? Assuming non-collusion (independence) how do retries affect our statistics, especially around false positives / false negatives ? What would it take to get us to, say, 99.99%[1] confidence that we're selecting a non-tampered video?

[1] Arbitrary value for now, but it'd be interesting to have a confidence number for a given number of retries.

**Comment by ndujar** [comment]
> Agreed that the enforcement of the policy belongs on the broadcaster. But it'd be good to know what appropriate default values are for this classifier, since the policy for a given verifier is likely to be tuned to the characteristics of the verifier itself.

Currently the only parameter we pass to a verifier is actually the max_samples (i.e. sampling rate) which is rather arbitrarily set to 10 frames.
Then, different configurations of the model (supervised, unsupervised, choice of metrics,...) give different confusion matrices (TNR, FNR, TPR and FPR) and hence different values for the F20 accuracy metric.

> Perhaps we can start with a default of 100% in a first pass to give the broadcaster the best possible defense against tampered video and then adjust after observing some e2e performance?

Using a sample rate other 100% is not taking into account yet another dimension of the problem: the confidence threshold derived from the random sampling of the frames. If we were to execute the same API call 100 times for the same source and rendition, we may still find a number N of false negatives when the frame configuration happens to be adverse enough.
We need to carry away a deeper study in this regard to obtain those confidence thresholds for TNR, FNR, TPR and FPR. 
This is not really a daunting task and would give us a corresponding confidence threshold for a given sample rate, which would enable us to define better the value of the retry parameter.
So, yes, I agree with @yondonfu that 100% is a good starting point for the sampling rate parameter :). This however, needs to be rearranged in the code because currently the number is not a ratio but an absolute number.

> What about starting with 0? The verifier API currently returns a tamper score (via the tamper field in the response) that will be negative if the result is classified as tamper and a lower score should represent a higher certainty that the result was tampered. So, a threshold set to 0 would mean that all video that is not classified as tampered would be accepted. If N results are below 0 (where N is the max # of retries), we pick the highest scoring result meaning we pick the result that is classified as the least distorted and that should be the least likely to be tampered [1].

I agree with @yondonfu that 0 is a reasonable start. We have to distinguish the two cases:
* Positive values (low or high), where the value is not telling us anything.
* Low negative values, where the magnitude may tell us the likelihood of being a false negative as @yondonfu  explains.
* High negative values, where the magnitude means larger amount of distorsion.

![image](https://user-images.githubusercontent.com/18640304/65258701-a6b86d00-db03-11e9-947e-0ac0da87ee14.png)

In the picture above, it is possible to observe the histogram for legitimate encoded assets and how those of 1080p are actually the ones having the most specimens closest to the decision boundary.
Given the specifications of the model, 1% of them have to fall outside (i.e. have negative values, i.e. be false negatives).
The effect of moving the threshold towards negative, say, putting it on -0.1, affects to the values of the TPR but also of the TNR, as we would be embracing more of the attacks (not represented in the picture).
Good news is, most of those attacks currently in the boundary correspond to small watermarks and low bitrates.

An elegant solution would be to return, for positive values, the SSIM and PSNR instead of the distance to the decision function. Given that we already have decoded and processed the frames this should be quite fast to compute. Besides, these metrics are meant to work well in non-tampered environments, which is where we will be.

**Comment by j0sh** [comment]
Thanks for the writeup @ndujar - this is very helpful!

Just wanted to clarify a point of communication around what's meant by the "sampling rate".  In terms of policy-defined sampling, I was thinking the sampling rate would be the frequency at which we pass whole segments in to the verifier. Maybe we should have a different term for this such as "verifier invocation rate".

We could expand this definition to include the "internal parameterization" of a particular verifier, such as the classifier's 10-frame sampling rate, but I'm leaning towards leaving those internal parameters fixed for the same reasons we want to set the [model once](https://github.com/livepeer/verification-classifier/issues/63) at start-up time. Totally open to discussing this point though.

That being said:

> Currently the only parameter we pass to a verifier is actually the max_samples (i.e. sampling rate) which is rather arbitrarily set to 10 frames. ... the confidence threshold derived from the random sampling of the frames

I'm OK leaving this up to your best judgement. Correct me if I'm incorrect, but I recall that increasing the sampling rate did not have a substantial impact on the F20 accuracy?

I understand that the classification of the actual frames is fairly quick. If we would still like to minimize the number of frames used during classification, then perhaps it might help to consider ways to make this sampling reproducible. One thought is the following:

1. Suppose we have a 256-bit hash function `H`, with a `maxValue=2^256-1`
2. Take a random seed `R`. The orchestrator should not know `R` in advance, otherwise they would be able to manipulate non-sampled frames. This can be generated once per source.
3. Take a hash of the source segment, `H(S)`
4. For each frame at index `i`, take ` sampleValue = H(H(S) | i | R) / maxValue`
5. `If ( sampleValue < samplingRate) classifyFrame()`

In order to re-run the verification to obtain the same results, the random seed `R` needs to be supplied to any subsequent verification run. There are ways to deterministically generate `R` while maintaining secrecy from the orchestrator, but not sure if that's needed here.

This method ensures that the same frames are picked for each verification run, even across renditions, although the actual number of frames chosen may vary due to the framerate, etc.

(It would be faster to run H(S) over the compressed data, rather than decoded data.)

> Positive values (low or high), where the value is not telling us anything.
> An elegant solution would be to return, for positive values, the SSIM and PSNR

Curious: does this imply there is basically zero correlation between the magnitude of a positive "tamper" score and the SSIM/PSNR score?

We can certainly return SSIM/PSNR scores if you believe that will help. Just noting that those are per-frame metrics, so we'd need to consider how to generate a final score out of that. Some more discussion about a hypothetical SSIM verifier here, would love your input! https://github.com/livepeer/go-livepeer/issues/1093#issuecomment-532800449

**Comment by ndujar** [comment]
> Just wanted to clarify a point of communication around what's meant by the "sampling rate". In terms of policy-defined sampling, I was thinking the sampling rate would be the frequency at which we pass whole segments in to the verifier. 

Oh, that's correct. We mentioned this in other conversations. I got confused. Thanks @j0sh for the clarification.

>Maybe we should have a different term for this such as "verifier invocation rate".

I believe this may be more accurate, but now that it is clarified, its just a matter of context.
Besides, we might as well rename the "sampling rate" we use for the frames with something more accurate as well, instead of frame_num.

>I'm OK leaving this up to your best judgement. Correct me if I'm incorrect, but I recall that increasing the sampling rate did not have a substantial impact on the F20 accuracy?
When 100% of the frames are used for the metric, the confidence with which that F20 is accounted for is 100%.
In the experiments we did, we only took in consideration a single invocation for measuring that F20.  However, as we reduce the number of sampled frames, the probability of the verifier to misclassify an attack raises, so the values of F20 indeed are similar but for the sake of correction they have to be accompanied with a confidence threshold. This threshold needs to be to calculated by realizing some experiments. We noticed this aspect after running some tests for the API. After 100 executions, in 4-5 requests the same true positive videos were regarded as false negative by the classifier. As we increase the number of frames, this confidence threshold raises.

**Comment by ndujar** [comment]
>I understand that the classification of the actual frames is fairly quick. If we would still like to minimize the number of frames used during classification, then perhaps it might help to consider ways to make this sampling reproducible. One thought is the following:
    Suppose we have a 256-bit hash function H, with a maxValue=2^256-1
    Take a random seed R. The orchestrator should not know R in advance, otherwise they would be able to manipulate non-sampled frames. This can be generated once per source.
    Take a hash of the source segment, H(S)
    For each frame at index i, take sampleValue = H(H(S) | i | R) / maxValue
    If ( sampleValue < samplingRate) classifyFrame()
In order to re-run the verification to obtain the same results, the random seed R needs to be supplied to any subsequent verification run. There are ways to deterministically generate R while maintaining secrecy from the orchestrator, but not sure if that's needed here.
This method ensures that the same frames are picked for each verification run, even across renditions, although the actual number of frames chosen may vary due to the framerate, etc.
(It would be faster to run H(S) over the compressed data, rather than decoded data.)

Currently the generation of the list of sampled frames is done internally in the initialization of the verifier, and remains fixed for all the renditions during the verification.
If I understand well, the proposal involves the supply of a list externally generated in order to enable reproducibility.
How about adding the internally generated list to the logs? Don't you think it would be easier to implement? Or am I being overly simplistic here?

**Comment by ndujar** [comment]
>We can certainly return SSIM/PSNR scores if you believe that will help. Just noting that those are per-frame metrics, so we'd need to consider how to generate a final score out of that. Some more discussion about a hypothetical SSIM verifier here, would love your input! https://github.com/livepeer/go-livepeer/issues/1093#issuecomment-532800449

Indeed this was part of the roadmap: once tamper verification is achieved we can be certain that we are dealing with transcodifications not subject to attacks.
The idea is that then a quality metric will not only help us discriminate further those false positives but also give a rank to the true positives.
In this sense, I believe that one concern will be performance of the metrics computation, but we have already done some progress at this point and found the main bottlenecks.

**Comment by j0sh** [comment]
> However, as we reduce the number of sampled frames, the probability of the verifier to misclassify an attack raises, so the values of F20 indeed are similar but for the sake of correction they have to be accompanied with a confidence threshold. This threshold needs to be to calculated by realizing some experiments.

Looking forward to this! Will this confidence score be fixed / derived from the frame sampling rate, or will it vary based on the results of a particular classifier run?

> How about adding the internally generated list to the logs? Don't you think it would be easier to implement? Or am I being overly simplistic here?

You are correct, that is a much better method - it is me that's being overly complicated :)

**Comment by j0sh** [comment]
>>>  If we would still like to minimize the number of frames used during classification, then perhaps it might help to consider ways to make this sampling reproducible
>>
>> How about adding the internally generated list to the logs? Don't you think it would be easier to implement? Or am I being overly simplistic here?
>
> You are correct, that is a much better method - it is me that's being overly complicated :)

One additional thought about this. Returning a list would work nicely in the case where we're retrying verification sequentially: for a retry, we can just pass in the list from the previous verification in order to check the same sequence of frames.

However, if we're doing parallel transcodes in order to minimize retry latency, then we might want some way to also perform verifications in parallel without waiting for the very first pass to complete. Having a stateless method to reproduce the sampling sequence would allow us to run parallel verifications over the same sequence of frames.

An alternative to this, of course, is just to sample different frames for each transcoded segment, but I imagine this is also subject to the confidence-interval analysis.

**Comment by ndujar** [comment]
> One additional thought about this. Returning a list would work nicely in the case where we're retrying verification sequentially: for a retry, we can just pass in the list from the previous verification in order to check the same sequence of frames.
However, if we're doing parallel transcodes in order to minimize retry latency, then we might want some way to also perform verifications in parallel without waiting for the very first pass to complete. Having a stateless method to reproduce the sampling sequence would allow us to run parallel verifications over the same sequence of frames.
An alternative to this, of course, is just to sample different frames for each transcoded segment, but I imagine this is also subject to the confidence-interval analysis.

I don't think I completely understand the concept of sequential verification. 
As I imagine the parallel verification process is as follows:
* Broadcaster proposes a job
* Several Orchestrators take the job
* Jobs get done asynchronously
* Verifier is triggered by a 'job completion event'
  * Option A: Verifier waits until all Orchestrators have their jobs sent or a timeout, storing the list of URLs and then running and sending all results together to Broadcaster
  * Option B: Verifier is triggered on each job completion, sending the results to the the Broadcaster
* Broadcaster chooses best job based on highest SSIM/PSNR combination

Is perhaps Option B what is meant by running verification sequentially?

**Comment by ndujar** [comment]
>   However, as we reduce the number of sampled frames, the probability of the verifier to misclassify an attack raises, so the values of F20 indeed are similar but for the sake of correction they have to be accompanied with a confidence threshold. This threshold needs to be to calculated by realizing some experiments.
Looking forward to this! Will this confidence score be fixed / derived from the frame sampling rate, or will it vary based on the results of a particular classifier run?

I have created an issue describing the process [here](https://app.zenhub.com/workspaces/transcoding-verification-5ce878fbf542de6fd8ff749b/issues/livepeer/verification-classifier/67).
Basically, the confidence is derived from the frame sampling rate, after several thousand runs over the classifier.

**Comment by j0sh** [comment]
> Is perhaps Option B what is meant by running verification sequentially?

Yes - Option B was what I was thinking for both parallel and sequential verification -  the verifier would be invoked immediately after receiving the segment. With sequential verification, we retry until a quality target is hit, or we run out of retries:

```
func job(transcoder) {
  score = verify(transcode(transcoder, segment))
  return score
}

transcoders = [ ... ] // list of transcoders
score = retries = 0
while score < target && retries < maxRetries {
  score = job(transcoders[retries])
  retries++
}
```

Parallel verification is as you described, which runs all the `jobs` at once and gathers the results at the end, where the verification is invoked immediately after receiving the transcoded results:

```
scores = parallel_map(job, transcoders)
best = max(scores)
```

The process could be 'short circuited' here to exit early once a quality threshold has been hit, so we don't need to wait for all responses.

> Option A: Verifier waits until all Orchestrators have their jobs sent or a timeout, storing the list of URLs and then running and sending all results together to Broadcaster

Option A is an interesting alternative - gather all the transcoded results, then submit them all to the verifier at once, as a group. The only downside here I can see is there's no opportunity to short-circuit the process, since we need to wait for all transcoded results prior to checking the scores.

**Comment by j0sh** [comment]
We can probably close this now since https://github.com/livepeer/go-livepeer/pull/1096 is merged so we do have a certain set of "defaults" albeit hard-coded; not much about the behavior is configurable.
